Add _get_type() API to RRef

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44716 Add bound method tests for async_execution with RRef helper
* #44666 Make async_execution compatible with RRef helpers
* **#44663 Add _get_type() API to RRef**

The new API returns the type of the data object referenced by this
`RRef`. On the owner, this is same as `type(rref.local_value())`.
On a user, this will trigger an RPC to fetch the `type` object from
the owner. After this function is run once, the `type` object is
cached by the `RRef`, and subsequent invocations no longer trigger
RPC.

closes #33210

Differential Revision: [D23691990](https://our.internmc.facebook.com/intern/diff/D23691990)

mrshenli committed:
Add _get_type() API to RRef

The new API returns the type of the data object referenced by this
`RRef`. On the owner, this is same as `type(rref.local_value())`.
On a user, this will trigger an RPC to fetch the `type` object from
the owner. After this function is run once, the `type` object is
cached by the `RRef`, and subsequent invocations no longer trigger
RPC.

closes #33210

[ghstack-poisoned]

mrshenli committed:
Update on "Add _get_type() API to RRef"


The new API returns the type of the data object referenced by this
`RRef`. On the owner, this is same as `type(rref.local_value())`.
On a user, this will trigger an RPC to fetch the `type` object from
the owner. After this function is run once, the `type` object is
cached by the `RRef`, and subsequent invocations no longer trigger
RPC.

closes #33210

[ghstack-poisoned]

mrshenli committed:
Update on "Add _get_type() API to RRef"


The new API returns the type of the data object referenced by this
`RRef`. On the owner, this is same as `type(rref.local_value())`.
On a user, this will trigger an RPC to fetch the `type` object from
the owner. After this function is run once, the `type` object is
cached by the `RRef`, and subsequent invocations no longer trigger
RPC.

closes #33210

Differential Revision: [D23691990](https://our.internmc.facebook.com/intern/diff/D23691990)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1985,6 +1985,51 @@ def test_return_local_rrefs(self):
+        with torch.autograd.profiler.profile() as p:
+            t = rref._get_type()
+
+        print(p.function_events)
```
Nit: Remove print before landing

mrshenli replied:
ah, my bad. Thanks for spotting this!

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1985,6 +1985,51 @@ def test_return_local_rrefs(self):
+
+        with torch.autograd.profiler.profile() as p:
+            for _ in range(10):
+                t = rref._get_type()
```
nit: t is not used?

mrshenli replied:
I should have added another `self.assertEqual(t, type(torch.ones(2)))` to verify that the type is still the same.

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1985,6 +1985,51 @@ def test_return_local_rrefs(self):
+    @dist_init
+    def test_rref_type_slow_init(self):
+        dst = worker_name((self.rank + 1) % self.world_size)
+        rref = rpc.remote(dst, RpcTest._slow_add, args=(torch.ones(2), 1))
```
What happens if `rpc.remote()` call times out and then the user calls `_get_type()`? Will we just handle this error situation by the `_get_type()` RPC also timing out? Would we want to eventually error out in an "eager" way if we know `rpc.remote()` for the RRef we're trying to get the type of has already failed?

rohan-varma replied:
There is `getTimedOut()` on RRefs, maybe we can use/generalize this if we want to do the above.

mrshenli replied:
It should have the same behavior as calling normal RPC on a timed-out RRef. Let me add a test to check the error message.

mrshenli replied:
Tried the a similar test as `test_rref_timeout`, it will wait until the RRef is ready, even if the `remote` function timed out. Let me create an issue to track.

mrshenli replied:
#44771

rohan-varma reviewed (approved):
LGTM

mrshenli committed:
Update on "Add _get_type() API to RRef"


The new API returns the type of the data object referenced by this
`RRef`. On the owner, this is same as `type(rref.local_value())`.
On a user, this will trigger an RPC to fetch the `type` object from
the owner. After this function is run once, the `type` object is
cached by the `RRef`, and subsequent invocations no longer trigger
RPC.

closes #33210

Differential Revision: [D23691990](https://our.internmc.facebook.com/intern/diff/D23691990)

[ghstack-poisoned]
